frontend/src/components/Layout/AppBar/Buttons/TaskManagerButton.vue

```diff
@@ -70,104 +68,57 @@ const props = withDefaults(
 );
 
 const menu = ref(false);
-const taskList = ref<TaskInfo[]>([]);
-const scheduledTimeout = ref(0);
 const taskManager = taskManagerStore();
-
-const buttonColor = computed(() => {
-  return taskList.value.every((task) => {
-    return task.progress === 100;
-  })
-    ? 'success'
-    : undefined;
-});
-
-const showButton = computed(() => taskList.value.length > 0);
-
-/**
- *  Clear the tasks list
- */
-function clearState(): void {
-  taskList.value = [];
-}
-
-/**
- * Set a timeout to hide the button
- */
-function setTimeout(): void {
-  scheduledTimeout.value = window.setTimeout(clearState, props.timeout);
-}
-
-/**
- * Locally retrieve and parse the tasks list
- */
-function getTaskList(): void {
-  const list: Array<TaskInfo> = [];
-
-  for (const task of taskManager.tasks) {
-    switch (task.type) {
+let frozenTaskList: readonly TaskInfo[] = [];
+const runningTaskList = computed<TaskInfo[]>(() => {
+  return taskManager.tasks.map((t) => {
+    switch (t.type) {
       case TaskType.ConfigSync: {
-        list.push({
-          progress: undefined,
+        return {
+          progress: t.progress,
           textKey: 'appbar.tasks.configSync',
-          id: task.id
-        });
-
-        break;
+          id: t.id
+        };
       }
       case TaskType.LibraryRefresh: {
-        list.push({
-          progress: task.progress,
+        return {
+          progress: t.progress,
           textKey: 'appbar.tasks.scanningLibrary',
           textParams: {
-            library: task.data || ''
+            library: t.data || ''
           },
-          id: task.id
-        });
-
-        break;
+          id: t.id
+        };
       }
     }
-  }
+  });
+});
 
-  const taskIds = new Set(
-    list.map((task) => {
-      return task.id;
-    })
-  );
-  const finishedTasks: Array<TaskInfo> = [];
+const allCompleted = computed(() =>
+  runningTaskList.value.every((t) => t.progress === 100)
+);
 
-  for (const task of taskList.value) {
-    if (!taskIds.has(task.id)) {
-      task.progress = 100;
-      finishedTasks.push(task);
-    }
-  }
+const buttonColor = computed(() => {
+  return allCompleted.value ? 'success' : undefined;
+});
 
-  taskList.value = [...list, ...finishedTasks];
-}
+const taskList = computed(() => {
+  return menu.value && allCompleted.value
```

Comment:
@aweeebs The idea is to have this behaviour:

* If at least one task is running (the spinner is visible in the button) and the menu is open: the task list is fully reactive with the store (which removes completed tasks after 5 seconds, the time we have configured right now). *Completed tasks will disappear in front of the user in this situation*.

* If all the tasks are completed and the menu is closed: a tick is shown in the button and the task list of the component is tied to the store, so the button disappears 5 seconds after the latest finished task.

* If all the tasks are completed and the menu is open: if we tie the task list to the state of the store, *the menu will be closed in front of the user 5 seconds after the last completed task*. That's bad UX and we assume that if an user opens a menu, it should stay open because he/she wants to see the piece of information it shows.
That's why the ``taskList`` switches between the non reactive array made in the watcher and the computed ``runningTaskList`` if the menu is open and all the tasks are completed.

Reply:
Thanks for enumerating those scenarios, that's very helpful. I have one follow up question: if the menu is open and tasks have completed, should they still disappear after the configured timeout or all completed tasks expected to stay around in that case until the menu is closed?

Based on the answer I have some proposed changes :D

Reply:
@ferferga as a followup to the above, it would be good to enumerate the requirements of how we expect the component to work in a docblock comment in this file as well!

Reply:
One last note about why I ask. The current behavior is still surprising to me, since the non-reactive list will only keep items in it as long as the existing task list contains non-completed items. These will instantly disappear if a new incomplete task item is added in.

I think the requirements for this component can be distilled down to:
- The button should display if there are any tasks in the list OR the user already has the menu open
- A checkmark will be displayed if all tasks are completed
- Completed tasks should never disappear as long as the menu is open <-- the current behavior kind of implements this, but it's hacky and doesn't quite work in all scenarios (though those scenarios may not be possible in the UI today). Is this an actual requirement?

Reply:
@aweebs They should stay when the menu is open, otherwise the menu will be closed after the 5 second timeout clears all the tasks.

The 5 second timeout is for when you already have running tasks and to keep the button with the success checkmark. Without the "completed task" state, no checkmark will be ever shown.

What's exactly not working?

Reply:
Here's the situation where this won't work. Let's say that the menu started open and stays open for the duration of this scenario.
- Task A starts and completes. It is now on a 5 second timer before it disappears.
- Task B arrives before Task A's 5 seconds is up, and stays in progress for 30 seconds
- Because not all tasks are completed, Task A will disappear from the list since Task B is still in progress and thus not all tasks are completed so the reactive view will continue to be used

Reply:
If we're ok with the tasks disappearing from the list after 5 seconds even if the menu is open, we can change the `showButton` computed to be if `menu` is true and the reactive list contains anything.
```
const showButton = computed(() => menu.value || taskManager.tasks.length > 0);
```
This will keep the button available and the list open if the user has already opened it, but does mean that the list could be empty.

Reply:
@aweebs That's intended: we can just find a good balance between how much time we expect the user to find relevant that X task has been completed. So, while the menu is open, tasks are cleared up following the timeout so the menu doesn't get cluttered (but you're expected to check the details, as the menu is opened)

What you don't expect as an user is for the menu to close automatically. Making the non reactive list when all the items are completed avoids that.

Reply:
@aweebs Having an empty menu open is useless. *We strictly want to ensure that the user is always in the control of the opening state of the menu and that info its shown at all times while it's open*.

Reply:
@ferferga if the menu is empty we could simply show a message indicating that "There are no active tasks". From my recommended changes, this scenario would only happen if the user had the menu open and left it open while all tasks disappeared.

On the flip side, if we do want to show all completed tasks in the list as long as it stays open, we should fix the bug that I described by keeping a list of all completed tasks when the menu is open and merging it with the list of tasks that come from the store.

Reply:
Option (2) would look like
```
// a "cache" for completed tasks 
const completedTasks = ref<TaskThing>([]);

watch([menu, taskManager.tasks.length], () => {
  if (menu) {
    // the menu is open so make sure any completed tasks are kept in our cache
    completedTasks.push(...taskManager.tasks.filter((t) => /** filter out tasks that are not completed and don't already exist in `completedTasks`**/));
  } else {
    // the menu is closed so stop caching completed tasks
    completedTasks.value = [];
  }
}, { immediate: true });

const taskListForUiDisplay = computed(() => [...completedTasks, taskManager.tasks.filter((t) => !t.completed)]);

Reply:
Basically, I agree that the system you have right works well enough given how few tasks there are in the system, and the situations I've described generally won't happen. But I think it's very low effort to implement a solution which is more robust, has no weird edges, and has a much more understandable operation if/when future tasks are implemented. In addition to user expectations, I think it's also clearer to developers what the intended behaviour is since the edges cases I've pointed out won't be clear to a developer if the behavior is a bug or intended.

Reply:
@aweebs You're explaining me the *hows*, but what about the *whys*. Why exactly do you think the current approach is not correct?

In my opinion:

* Showing a "No tasks message" and an empty menu is the same thing. 
* When a task is running, what's relevant are the running task(s). Completed ones are a secondary concern. This changes when all the tasks are completed: what matters to the user starts to be the completed tasks.
* I inspired myself in how toasts in places like Google Drive work: when they are hovered, they will be kept onscreen. When they're not, they are expected to disappear. When something is being performed (like restoring files from trash), the running one is kept all the time, while the ones showing info are bound to the same timeout as the others.
In our example, change the mouse hover with the menu.

Reply:
@aweebs Your last message was sent when I was typing my last one.
What's the difference in regards how the tasks are shown when more tasks get into the system? I think the concept of completed/running task and percentages or indeterminate progress are expected to be permanent.

Reply:
@ferferga I agree with your reasoning for wanting to keep all completed tasks showing in the list as long as it is open.
For that scenario I'd push for what I described in option (2), since as long as the menu is open, it will continue to show any tasks that have completed since the menu was opened, and guarantees that the list will never be empty as long as it's visible because of that.

The main difference between it and your implementation is that your implementation has edges cases where completed items can disappear from the menu list while it's open (eg. when new tasks are spawned while the list is open). So a user might be hovering and ready to click on a completed task, but it could suddenly disappear. As mentioned above, this generally won't happen today since tasks aren't used across the system very much, but I think it's better to have a system which will work correctly in the future as more tasks are added or used.

Reply:
Replacing the logic to keep completed items around if the menu is open with logic that only does it if they're hovered also seems fine, though I think that's more susceptible to small issues down the road, like the user hovering over an option, then another item in the list disappearing, which would change the user's currently hovered item, and thus let the previous one they were hovering disappear potentially. More crazy edge cases that may or may not happen :D

Using the menu open as a trigger to keep things around is easier since new incoming tasks will always be shown at the end of the list and so we don't have to deal with the list items ever getting shorter as long as the menu is open.